lib/assessment.js

```diff
@@ -8,33 +8,87 @@
 // Question 1
 // Write a callback for reduce that selects
 // the minimum value in the array
-var min = function min() {
+var min = function min() {previousValue, currentValue, index, array)
+return previousValue < currentValue ? previousValue : currentValue;
+};
+
 
+// callback(previousValue, currentValue, index, array)
+var reduce = function reduce(array, callback, initialValue) {
+  var previousValue;
+  if (initialValue !== undefined) {
+    previousValue = initialValue;
+  } else {
+    array = array.slice();
+    previousValue = array.shift();
+  }
 };
 
+console.log(reduce(numbers, min));
+
 // Question 2
 // Write a callback for reduce that produces
 // the sum of the values in the array
-var sum = function sum() {
+var sum = function sum(previousValue, currentValue, index, array) {
+  return previousValue + currentValue;
+};
+
+var reduce = function reduce(array, callback, initialValue) {
+  var previousValue;
+  if (initialValue !== undefined) {
+    previousValue = initialValue;
+  } else {
+    array = array.slice();
+    previousValue = array.shift();
+  }
+
+  array.forEach(function(currentValue, index, array) {
+    previousValue = callback(previousValue, currentValue, index, array);
+  });
 
+  return previousValue;
 };
 
+var numbers = [1, 2, 3, 4, 5];
+
+console.log(reduce(numbers, sum));
+);
+
 // Question 3
 // Write a function that returns true
 // if invoking the callback for any element
 // in the array returns a truthy value
 var any = function any(array, callback, thisArg) {
+  var result = true;
+  array.forEach(function(currentValue, index, array) {
+    // if callback ever returns false, result will be set
+    // to false amd the callback will not be called in any
+    // subsequent iterations.
+    result = result || callback(currentValue, index, array);
+  });
 
+  return result;
 };
 
+
 // Question 4
 // Write a function that returns a new array
 // made up of the array elements for which invoking
 // the callback returns a truthy value
 var select = function select(array, callback, thisArg) {
+  var newArray = [];
```

Comment:
Here, we want to use the fact that our callback should return a boolean for each element we pass to it (if it doesn't, that means the user is wrong, not us!) to determine whether we should add that element to the new list we're building.